File: apps/apple/Shared/Fingerprint.swift
//
//  Fingerprint.swift
//  Empreinte de la clef maîtresse.
//
//  Une faute de frappe sur la clef ne se voit pas : elle produit simplement un
//  autre mot de passe, valide en apparence. On ne s'en aperçoit qu'au refus de
//  connexion, sans savoir si le tort vient de la clef, du site ou des réglages.
//
//  Source unique : shared/apple/Fingerprint.swift
//  Spécification : shared/spec/fingerprint.md
//

import CommonCrypto
import Foundation
import SwiftUI

public enum Fingerprint {

    private static let salt = "thecode-fingerprint/v1"
    private static let iterations: UInt32 = 600_000
    private static let keyBytes = 32

    /// Sans 0/O ni 1/I/L : une empreinte se lit parfois à voix haute, elle ne
    /// doit laisser aucune hésitation.
    private static let alphabet = Array("23456789ABCDEFGHJKMNPQRSTUVWXYZ")
    private static let length = 3

    /// Douze teintes distinctes, pour un repère visuel immédiat.
    private static let palette: [Color] = [
        Color(red: 0.898, green: 0.282, blue: 0.302),
        Color(red: 0.969, green: 0.420, blue: 0.082),
        Color(red: 1.000, green: 0.698, blue: 0.141),
        Color(red: 0.741, green: 0.933, blue: 0.388),
        Color(red: 0.275, green: 0.655, blue: 0.345),
        Color(red: 0.161, green: 0.639, blue: 0.514),
        Color(red: 0.000, green: 0.635, blue: 0.780),
        Color(red: 0.000, green: 0.565, blue: 1.000),
        Color(red: 0.243, green: 0.388, blue: 0.867),
        Color(red: 0.431, green: 0.337, blue: 0.812),
        Color(red: 0.839, green: 0.251, blue: 0.624),
        Color(red: 0.914, green: 0.239, blue: 0.510),
    ]

    public struct Result {
        public let text: String
        public let color: Color
    }

    private static func derive(_ masterKey: String) -> [UInt8]? {
        var output = [UInt8](repeating: 0, count: keyBytes)
        let saltBytes = Array(salt.utf8)

        let status = masterKey.withCString { keyPointer in
            CCKeyDerivationPBKDF(
                CCPBKDFAlgorithm(kCCPBKDF2),
                keyPointer,
                strlen(keyPointer),
                saltBytes,
                saltBytes.count,
                CCPseudoRandomAlgorithm(kCCPRFHmacAlgSHA256),
                iterations,
                &output,
                keyBytes
            )
        }

        // Plutôt qu'afficher une empreinte fausse, on n'en affiche aucune : un
        // repère erroné serait pire que pas de repère du tout.
        return status == kCCSuccess ? output : nil
    }

    /// Empreinte de la clef, ou `nil` si elle ne peut pas être calculée.
    public static func of(_ masterKey: String) -> Result? {
        guard !masterKey.isEmpty, let raw = derive(masterKey) else { return nil }

        let text = (0..<length)
            .map { alphabet[Int(raw[$0]) % alphabet.count] }
            .reduce(into: "") { $0.append($1) }

        return Result(text: text, color: palette[Int(raw[length]) % palette.count])
    }
}

File: apps/android/app/src/main/java/fr/juliette/thecode/Fingerprint.java
package fr.juliette.thecode;

import androidx.annotation.NonNull;

import java.nio.charset.StandardCharsets;
import java.security.NoSuchAlgorithmException;
import java.security.spec.InvalidKeySpecException;

import javax.crypto.SecretKeyFactory;
import javax.crypto.spec.PBEKeySpec;

/**
 * Empreinte de la clef maîtresse.
 *
 * Une faute de frappe sur la clef ne se voit pas : elle produit simplement un
 * autre mot de passe, valide en apparence. On ne s'en aperçoit qu'au refus de
 * connexion, sans savoir si le tort vient de la clef, du site ou des réglages.
 *
 * Spécification : shared/spec/fingerprint.md
 */
public final class Fingerprint {

    private static final byte[] SALT = "thecode-fingerprint/v1".getBytes(StandardCharsets.UTF_8);
    private static final int ITERATIONS = 600_000;
    private static final int KEY_BITS = 256;

    /**
     * Sans 0/O ni 1/I/L : une empreinte se lit parfois à voix haute, elle ne
     * doit laisser aucune hésitation.
     */
    private static final String ALPHABET = "23456789ABCDEFGHJKMNPQRSTUVWXYZ";
    private static final int LENGTH = 3;

    /** Douze teintes distinctes, pour un repère visuel immédiat. */
    private static final int[] COLORS = {
            0xFFE5484D, 0xFFF76B15, 0xFFFFB224, 0xFFBDEE63,
            0xFF46A758, 0xFF29A383, 0xFF00A2C7, 0xFF0090FF,
            0xFF3E63DD, 0xFF6E56CF, 0xFFD6409F, 0xFFE93D82,
    };

    private Fingerprint() {}

    /** Résultat : trois caractères et une couleur. */
    public static final class Result {
        public final String text;
        public final int color;

        Result(String text, int color) {
            this.text = text;
            this.color = color;
        }
    }

    private static byte[] derive(String masterKey) {
        try {
            PBEKeySpec spec = new PBEKeySpec(
                    masterKey.toCharArray(), SALT, ITERATIONS, KEY_BITS);
            return SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256")
                    .generateSecret(spec).getEncoded();
        } catch (NoSuchAlgorithmException | InvalidKeySpecException e) {
            // Algorithme absent : plutôt que d'afficher une empreinte fausse,
            // on n'en affiche aucune. Un repère erroné serait pire que pas de
            // repère du tout.
            return null;
        }
    }

    /** Empreinte de la clef, ou {@code null} si elle ne peut pas être calculée. */
    public static Result of(@NonNull String masterKey) {
        if (masterKey.isEmpty()) return null;

        byte[] raw = derive(masterKey);
        if (raw == null) return null;

        StringBuilder text = new StringBuilder(LENGTH);
        for (int i = 0; i < LENGTH; i++) {
            text.append(ALPHABET.charAt((raw[i] & 0xFF) % ALPHABET.length()));
        }
        return new Result(text.toString(), COLORS[(raw[LENGTH] & 0xFF) % COLORS.length]);
    }
}
